EntryProcessorOffloadableTest.testHeartBeatsComingWhenEntryPropcessorOffloaded failure

*master*

Failed on Oracle JDK6:
https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-OracleJDK1.6-freestyle/lastCompletedBuild/testReport/

Stacktrace:
```
java.lang.AssertionError: Heartbeats came too close to each other: 773ms
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.map.EntryProcessorOffloadableTest.testHeartBeatsComingWhenEntryPropcessorOffloaded(EntryProcessorOffloadableTest.java:998)
```

Standard output:
```
04:28:20, accumulated pauses: 860 ms, max pause: 659 ms, pauses over 1000 ms: 0
04:28:25, accumulated pauses: 3889 ms, max pause: 3032 ms, pauses over 1000 ms: 1
04:28:30, accumulated pauses: 246 ms, max pause: 190 ms, pauses over 1000 ms: 0
```

Maybe related to https://github.com/hazelcast/hazelcast-enterprise/issues/2012 @tkountis  ?

*maintenance* (commit 56dd06f989e32e753c3d98a3b47faae259eec1c6)

Reopening as it failed with Oracle JDK6, but this time with a different stacktrace: https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.maintenance-OracleJDK6/lastCompletedBuild/testReport/com.hazelcast.map/EntryProcessorOffloadableTest/testHeartBeatsComingWhenEntryPropcessorOffloaded_0__BINARY_sync_0_async_0_/

Stacktrace:
```
java.lang.AssertionError: At least 4 heartBeats expected, 3 came.
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.map.EntryProcessorOffloadableTest.testHeartBeatsComingWhenEntryPropcessorOffloaded(EntryProcessorOffloadableTest.java:990)
```

Standard output:
```
16:58:20, accumulated pauses: 4229 ms, max pause: 1919 ms, pauses over 1000 ms: 2
16:58:25, accumulated pauses: 187 ms, max pause: 18 ms, pauses over 1000 ms: 0
16:58:30, accumulated pauses: 405 ms, max pause: 372 ms, pauses over 1000 ms: 0
```

--- a/hazelcast/src/test/java/com/hazelcast/map/EntryProcessorOffloadableTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/map/EntryProcessorOffloadableTest.java
@@ -47,6 +47,7 @@ import java.io.Serializable;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.HashSet;
+import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
@@ -54,7 +55,6 @@ import java.util.concurrent.Callable;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
-import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicLong;
 
 import static com.hazelcast.config.InMemoryFormat.BINARY;
@@ -952,13 +952,13 @@ public class EntryProcessorOffloadableTest extends HazelcastTestSupport {
     @Test
     public void testHeartBeatsComingWhenEntryPropcessorOffloaded() throws Exception {
         final String key = generateKeyOwnedBy(instances[1]);
-        SimpleValue givenValue = new SimpleValue(1);
+        TimestampedSimpleValue givenValue = new TimestampedSimpleValue(1);
 
         final IMap<Object, Object> map = instances[0].getMap(MAP_NAME);
         map.put(key, givenValue);
 
         final Address instance1Address = instances[1].getCluster().getLocalMember().getAddress();
-        final AtomicInteger heartBeatsCounter = new AtomicInteger(0);
+        final List<Long> heartBeatTimestamps = new LinkedList<Long>();
         Thread hbMonitorThread = new Thread() {
             public void run() {
                 NodeEngine nodeEngine = HazelcastTestSupport.getNodeEngineImpl(instances[0]);
@@ -971,28 +971,37 @@ public class EntryProcessorOffloadableTest extends HazelcastTestSupport {
                         long newlastbeat = timestamp.get();
                         if (lastbeat != newlastbeat) {
                             lastbeat = newlastbeat;
-                            heartBeatsCounter.incrementAndGet();
+                            heartBeatTimestamps.add(newlastbeat);
                         }
                     }
                     HazelcastTestSupport.sleepMillis(100);
                 }
             }
         };
+
         final int secondsToRun = 8;
-        final int expectedHeartBeats = secondsToRun / HEARTBEATS_INTERVAL_SEC;
         try {
             hbMonitorThread.start();
             map.executeOnKey(key, new TimeConsumingOffloadableTask(secondsToRun));
         } finally {
             hbMonitorThread.interrupt();
         }
-        int hbCount = heartBeatsCounter.get();
-        assertTrue("At least " + expectedHeartBeats + " heartBeats expected, " + hbCount + " came.",
-                hbCount >= expectedHeartBeats);
+
+        int heartBeatCount = 0;
+        TimestampedSimpleValue updatedValue = (TimestampedSimpleValue) map.get(key);
+        for (long heartBeatTimestamp : heartBeatTimestamps) {
+            if (heartBeatTimestamp > updatedValue.processStart
+                    && heartBeatTimestamp < updatedValue.processEnd) {
+                heartBeatCount++;
+            }
+        }
+
+        assertTrue("Heartbeats should be received while offloadable entry processor is running", heartBeatCount > 0);
     }
 
-    private static class TimeConsumingOffloadableTask implements EntryProcessor<String, SimpleValue>, Offloadable,
-            EntryBackupProcessor<String, SimpleValue>, Serializable {
+    private static class TimeConsumingOffloadableTask implements EntryProcessor<String, TimestampedSimpleValue>, Offloadable,
+                                                                 EntryBackupProcessor<String, TimestampedSimpleValue>,
+                                                                 Serializable {
 
         private final int secondsToWork;
 
@@ -1001,24 +1010,28 @@ public class EntryProcessorOffloadableTest extends HazelcastTestSupport {
         }
 
         @Override
-        public Object process(final Map.Entry<String, SimpleValue> entry) {
-            final SimpleValue value = entry.getValue();
+        public Object process(final Map.Entry<String, TimestampedSimpleValue> entry) {
+            final TimestampedSimpleValue value = entry.getValue();
+            value.processStart = System.currentTimeMillis();
             long endTime = TimeUnit.SECONDS.toMillis(secondsToWork) + System.currentTimeMillis() + 500L;
             do {
                 HazelcastTestSupport.sleepMillis(200);
                 value.i++;
                 entry.setValue(value);
             } while (System.currentTimeMillis() < endTime);
+            value.i = 1;
+            entry.setValue(value);
+            value.processEnd = System.currentTimeMillis();
             return null;
         }
 
         @Override
-        public EntryBackupProcessor<String, SimpleValue> getBackupProcessor() {
+        public EntryBackupProcessor<String, TimestampedSimpleValue> getBackupProcessor() {
             return this;
         }
 
         @Override
-        public void processBackup(Map.Entry<String, SimpleValue> entry) {
+        public void processBackup(Map.Entry<String, TimestampedSimpleValue> entry) {
             process(entry);
         }
 
@@ -1031,7 +1044,7 @@ public class EntryProcessorOffloadableTest extends HazelcastTestSupport {
 
     private static class SimpleValue implements Serializable {
 
-        public int i;
+        int i;
 
         SimpleValue() {
         }
@@ -1064,5 +1077,17 @@ public class EntryProcessorOffloadableTest extends HazelcastTestSupport {
         }
     }
 
+    private static class TimestampedSimpleValue extends SimpleValue {
+        private long processStart;
+        private long processEnd;
+
+        public TimestampedSimpleValue() {
+            super();
+        }
+
+        public TimestampedSimpleValue(int i) {
+            super(i);
+        }
+    }
 
 }

